torch_geometric/nn/dense/hoscpool.py

```diff
@@ -0,0 +1,153 @@
+
+        \mathcal{L}_o &= \frac{1}{\sqrt{K}-1} \bigg( \sqrt{K} - \frac{1}{\sqrt{N}}\sum_{j=1}^K ||S_{*j}||_F\bigg)
+
+    Args:
+        x (Tensor): Node feature tensor :math:`\mathbf{X} \in \mathbb{R}^{B
+            \times N \times F}` with batch-size :math:`B`, (maximum)
+            number of nodes :math:`N` for each graph, and feature dimension
+            :math:`F`.
+        adj (Tensor): adjacency matrix :math:`\mathbf{A} \in \mathbb{R}^{B \times N \times N}`.
+        s (Tensor): the learnable cluster assignment matrix :math:`\mathbf{S} \in \mathbb{R}^{B
+            \times N \times K}` with number of clusters :math:`K`. The softmax
+            does not have to be applied beforehand, since it is executed
+            within this method.
+        mu (Tensor, optional): scalar that controls the importance given to regularization loss
+        alpha (Tensor, optional): scalar in [0,1] controlling the importance granted
+            to higher-order information (in loss function).
+        new_ortho (BoolTensor, optional): either to use new proposed loss or old one
+        mask (BoolTensor, optional): Mask matrix
+            :math:`\mathbf{M} \in {\{ 0, 1 \}}^{B \times N}` indicating
+            the valid nodes for each graph. (default: :obj:`None`)
+
+    :rtype: (:class:`Tensor`, :class:`Tensor`, :class:`Tensor`,
+        :class:`Tensor`)
+    """
+    x = x.unsqueeze(0) if x.dim() == 2 else x
+    adj = adj.unsqueeze(0) if adj.dim() == 2 else adj
+    s = s.unsqueeze(0) if s.dim() == 2 else s
+
+    (batch_size, num_nodes, _), k = x.size(), s.size(-1)
+
+    s = torch.softmax(s, dim=-1)
+
+    if mask is not None:
+        mask = mask.view(batch_size, num_nodes, 1).to(x.dtype)
+        x, s = x * mask, s * mask
+
+    # Output adjacency and feature matrices
+    out = torch.matmul(s.transpose(1, 2), x)
+    out_adj = torch.matmul(torch.matmul(s.transpose(1, 2), adj), s)
+
+    # Motif adj matrix - not sym. normalised
+    motif_adj = torch.mul(torch.matmul(adj, adj), adj)
+    motif_out_adj = torch.matmul(torch.matmul(s.transpose(1, 2), motif_adj), s)
+
+    mincut_loss = ho_mincut_loss = 0
+    # 1st order MinCUT loss
+    if alpha < 1:
+        diag_SAS = torch.einsum("ijj->ij", out_adj.clone())
+        d_flat = torch.einsum("ijk->ij", adj.clone())
+        d = _rank3_diag(d_flat)
+        sds = torch.matmul(torch.matmul(s.transpose(1, 2), d), s)
+        diag_SDS = torch.einsum("ijk->ij", sds) + EPS
+        mincut_loss = -torch.sum(diag_SAS / diag_SDS, axis=1)
+        mincut_loss = 1 / k * torch.mean(mincut_loss)
+
+    # Higher order cut
+    if alpha > 0:
+        diag_SAS = torch.einsum("ijj->ij", motif_out_adj)
+        d_flat = torch.einsum("ijk->ij", motif_adj)
+        d = _rank3_diag(d_flat)
+        diag_SDS = (torch.einsum(
+            "ijk->ij", torch.matmul(torch.matmul(s.transpose(1, 2), d), s)) +
+                    EPS)
+        ho_mincut_loss = -torch.sum(diag_SAS / diag_SDS, axis=1)
```

Comment:
There is soem duplicated logic here, this and the next line could  be shared between the two alpha conditions?